docs/rfc/0003-collaborative-objects.md

```diff
@@ -0,0 +1,421 @@
+|  |  |--<schema change 2>
+|  |--<change 1>
+|  |--<change 2>
+|--<object 2 ID>
+|  ...
+```
+
+We refer to a directory in this tree as an "object directory" and the tuple 
+`(manifest, [schema changes], [changes])` as an "object state".
+
+Objects are created with a unique identifier at creation time. This identifier
+is the name of the directory within the `collaborative_objects` tree where the
+data for this object is kept. Each time we make a change to the object we store
+the new change (which is a binary blob) in the object directory and create a
+new commit. We also store a manifest file which contains the name of the data
+type of this object. This file is never changed and is used to allow
+applications to enumerate collaborative objects of a particular type. Type
+names should be some kind of human readable identifier - similar to XML
+namespaces.
+
+Each object is also created with a schema, this is a JSON schema document which
+is itself encoded as an automerge document within the `schema` directory. This
+allows for changes to the schema (as long as they are backwards and forwards
+compatible) to be distributed as additional changes within the `<object
+ID>/schema` directory.
+
+### Viewing the tracking graph
+
+We need to load the state of objects which multiple people within the
+tracking graph are collaborating on. Given a particular object ID we can load
+all the changes from each remote we have replicated and merge them all to
+obtain the final state of the object. With many remotes this could become
+expensive and there are opportunities to do some indexing ahead of time. A more
+pressing concern is that we want to ensure that the schema is respected, more
+on this later.
+
+
+### Updating objects
+
+To make a change to an object we load the existing messages for an object. The 
+application developer provide us with the binary representation of the change
+to that object. We apply the change and ensure that the new object state still
+matches the object schema. At this point the state of the object may depend on
+many contributions from the tracking graph - not just the ones in our own view
+of the project. This is a natural point at which to compact the document using
+automerge's compacted document format. We bundle all the changes into one
+compacted change file and replace all the changes in our view of the object
+with this compacted change file - then create a new commit, update our signed
+refs, and announce.
+
+### Strange Perspectives
+
+This model introduces some counter-intuitive properties. For example, I might
+"create an issue" in a repository and anyone who is tracking me would see that
+issue, but people who are tracking the project but don't have me in their
+tracking graph will only see the issue if the maintainer replies to it. It's
+hard to see how you would do things like "link to an issue" under these
+constraints. This is inherent to the network model though, rather than being a
```

Comment:
As you say, it's inherent. Consider, however, the role of seed nodes in our model: modulo their local access policy, they act as supernodes for the social network. Ie. a libral seed (which auto-tracks every peer id it learns about) would converge to a globally consistent view.

This poses two problems:

1. One might wish to _interrogate_ a seed's view of the object, but not fetch 10k remotes to do so
2. A maintainer might wish to apply _some_ of the updates the seed has applied, but leave out others (e.g because they contain spam)

This seems to go against the proposal to store _all_ objects in a single `tree` object, even though tempting due to the consistency guarantees. Not that it is _impossible_ to perform "shallow" fetches, but it'd be considerably more complicated. Also consider that we can't tell what was included in a partial merge other than evaluating the automerge, unlike actual merge (multi-parent) commits.

Reply:
Yeah ibagree with these concerns, im working on what this will look like for a tree-per-object, possibly with a path element in the ref for the type name of the object. This would allow things like fetching just the state of one object, or the state of all objects of one type.

it does increase the size of the initial ref advertisment during upload pack

Reply:
> im working on what this will look like for a tree-per-object

cool, I'll ask again then

> it does increase the size of the initial ref advertisment during upload pack

ya, we'll have some scalability issues with that either way. git protocol v2 will help (`ref-prefix`), but I'm thinking that we may want to prefix-compress refs on the wire at some point.

Reply:
Alternatively, and what I have written in a draft at the moment, we could add a query parameter to radicle URNs which specifies what collaborative objects you want to fetch. If that query parameter is omitted then no refs for them are advertised. Obviously this is a bit of a hack - but given that collaborative objects are Radicle specific things anyway that feels acceptable.